pkg/controllers/deprovisioning/expiration.go

```diff
@@ -57,64 +66,48 @@ func (e *Expiration) SortCandidates(nodes []CandidateNode) []CandidateNode {
 }
 
 // ComputeCommand generates a deprovisioning command given deprovisionable nodes
-func (e *Expiration) ComputeCommand(ctx context.Context, attempt int, candidates ...CandidateNode) (Command, error) {
+func (e *Expiration) ComputeCommand(ctx context.Context, candidates ...CandidateNode) (Command, error) {
+	candidates = e.SortCandidates(candidates)
 	pdbs, err := NewPDBLimits(ctx, e.kubeClient)
 	if err != nil {
 		return Command{action: actionFailed}, fmt.Errorf("tracking PodDisruptionBudgets, %w", err)
 	}
-	// is this a node that we can terminate?  This check is meant to be fast so we can save the expense of simulated
-	// scheduling unless its really needed
-	if !canBeTerminated(candidates[attempt], pdbs) {
-		return Command{action: actionNotPossible}, nil
-	}
-
-	// Only expire one node at a time.
-	node := candidates[attempt]
-	// Check if we need to create any nodes.
-	newNodes, allPodsScheduled, err := simulateScheduling(ctx, e.kubeClient, e.cluster, e.provisioner, node)
-	if err != nil {
-		// if a candidate node is now deleting, just retry
-		if errors.Is(err, errCandidateNodeDeleting) {
-			return Command{action: actionDoNothing}, nil
+	for _, candidate := range candidates {
+		// is this a node that we can terminate?  This check is meant to be fast so we can save the expense of simulated
+		// scheduling unless its really needed
+		if !canBeTerminated(candidate, pdbs) {
+			continue
 		}
-		return Command{}, err
-	}
-	// Log when all pods can't schedule, as the command will get executed immediately.
-	if !allPodsScheduled {
-		logging.FromContext(ctx).Infof("continuing to expire node %s after scheduling simulation failed to schedule all pods", node.Name)
-	}
 
-	// were we able to schedule all the pods on the inflight nodes?
-	if len(newNodes) == 0 {
+		// Check if we need to create any nodes.
+		newNodes, allPodsScheduled, err := simulateScheduling(ctx, e.kubeClient, e.cluster, e.provisioner, candidate)
+		if err != nil {
+			// if a candidate node is now deleting, just retry
+			if errors.Is(err, errCandidateNodeDeleting) {
+				continue
+			}
+			return Command{action: actionFailed}, err
+		}
+		// Log when all pods can't schedule, as the command will get executed immediately.
+		if !allPodsScheduled {
+			logging.FromContext(ctx).Infof("Continuing to expire node %s after scheduling simulation failed to schedule all pods", candidate.Name)
+		}
+		logging.FromContext(ctx).Infof("Triggering termination for expired node after %s (+%s)",
+			time.Duration(ptr.Int64Value(candidates[0].provisioner.Spec.TTLSecondsUntilExpired))*time.Second, time.Since(getExpirationTime(candidates[0].Node, candidates[0].provisioner)))
+		// were we able to schedule all the pods on the inflight nodes?
+		if len(newNodes) == 0 {
```

Comment:
Can we collapse delete and replace by looking at the replacementNode length and knowing that no replacement nodes means that we are performing delete?

Reply:
Yeah logically, I think we can do this, although we'll need to then implement an actionDeprovision in addition to a actionDoNothing, actionFailure (debatable), actionRetry, and maybe actionUnknown. While I think it does make the code slightly cleaner, I feel it makes the construct expectations of the code, and makes the code more error prone without a declarative statement.

We could theoretically do (action, err), where (nil, nil) == DoNothing, (_, err) == Failed, and (non-nil, nil) == Green Light, but I'd like to do this in a separate PR if at all to make sure it translates well.